Handle attachments web socket notifications on patient flow page

nmajor25 opened the pull request:
Shortcut Story ID: [sc-56327] [sc-57309]

In this PR, these are being handled only in the action sidebar on the patient flow page.

Coinciding with the backend PR: https://github.com/RoundingWell/care-ops-backend/pull/8787.

- [x]  Support `AttachmentAdded` notifications
- [x]  Support `FileReplaced` notifications
- [x]  Support `FileRemoved` notifications
- [x] Show paperclip icon in action list items when file is added/removed via websocket

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Enhanced WebSocket communication for comments and attachments
  - Improved file management functionality with real-time updates

- **Bug Fixes**
  - Refined event handling for adding and removing attachments
  - Updated message processing for better resource tracking

- **Improvements**
  - Streamlined integration of WebSocket notifications
  - Added support for file replacement and removal operations

- **Testing**
  - Expanded test coverage for patient flow and file attachment scenarios

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nmajor25 committed:
Handle `AttachmentAdded` web socket notifications

nmajor25 committed:
Handle `FileReplaced` web socket notifications

nmajor25 committed:
Handle `FileRemoved` web socket notifications

nmajor25 commented on src/js/entities-service/entities/actions.js:
```diff
@@ -250,6 +250,14 @@ const _Model = BaseModel.extend({
+    });
+
+    this.set({ _files: newFilesRelationship });
+  },
```
The paperclip attachments icon on action list items is shown or hidden based on the length of `action.get('_files')`.

We're currently not updating that when a file is removed manually by a user. And therefore it doesn't work out of the box for the `FileRemoved` ws notification.

So i added this method to the action entity service that filters out the file from `_files` by `id`. And then sets a new `_files` relationship array for the action.

This allows for the paperclip icon to be hidden after attachment deletion.

nmajor25 replied:
If this is how we want to implement this, we would just need to call `action.removeAttachment(model)` in the action sidebar app when a file is manually deleted by a user.

nmajor25 replied:
Currently in the app, the `.paperclip` icon isn't shown when an attachment is added and isn't hidden when an attachment is removed manually by a user in the sidebar.

I created a separate story [here]( https://app.shortcut.com/roundingwell/story/58034/in-action-list-items-show-hide-paperclip-icon-when-an-attachment-is-manually-added-removed-via-action-sidebar), that I can tackle in a separate PR.

nmajor25 commented on src/js/apps/patients/sidebar/action-sidebar_app.js:
```diff
@@ -82,6 +83,11 @@ export default App.extend(extend({
+    this.attachments.add(model);
+
+    Radio.request('ws', 'add', model);
+  },
```
This `onAttachmentAdded()` function name is almost identical to the already existing `onAddAttachment()` in this file.

We should probably name this something else, like `OnWsAttachmentAdded()`.

paulfalgout replied:
The events should possibly be: `ws:add:attachment`   Then `onWsAddAttachment` similar to `state:verb:subject` `onStateVerbSubject`

nmajor25 replied:
Makes sense, updated to that 👍

nmajor25 commented on src/js/apps/patients/sidebar/action-sidebar_app.js:
```diff
@@ -238,12 +246,12 @@ export default App.extend(extend({
   removeSubscriptions() {
     // for when sidebar is closed before comments api request is finished
-    if (!this.comments) return;
+    if (!this.comments || !this.attachments) return;
```
This doesn't seem right to me. I don't know if there's a situation where only one of these is `undefined`. Since the `beforeStart()` requests are called together, right?

Is there another way of guarding against the sidebar being closed before `beforeStart()` can finish?

paulfalgout replied:
I think this might be an `isRunning` or `isLoading` test of some sort

nmajor25 replied:
I switched it to use `this.isRunning()` 👍

nmajor25 commented on src/js/apps/patients/sidebar/action-sidebar_app.js:
```diff
@@ -230,20 +236,24 @@ export default App.extend(extend({
     });
     attachment.upload(file);
 
+    Radio.request('ws', 'add', file);
```
`file` is going to be `undefined` here. I think we need to change `upload(file)` to a promise and then do something like:

```js
attachment.upload(file).then(file => {
  Radio.request('ws', 'add', file);
})
```

paulfalgout replied:
actually I think it's just `Radio.request('ws', 'add', attachment)`

nmajor25 replied:
Updated to this ☝️

nmajor25 committed:
Use `this.isRunning()` as check if action sidebar has been closed prior to `beforeStart()` finishing

To protect the `removeSubscriptions()` function from running before comments or attachments api data is retrieved for the action sidebar

nmajor25 committed:
Use correct data when adding ws after attachment is uploaded

nmajor25 committed:
Use `state:verb:subject` pattern for websocket trigger events
